### PR TITLE
[gui] add ffi for getting mount target from source

### PR DESCRIPTION
**Pull request by andrei-toterman**
This PR makes the logic of computing a mount target path from a source path available in the FFI library that we use from Dart. This is so that the GUI can show a preview of what a target path will be before actually sending the request to the daemon.

### CODE REVIEW

**Review by levkropp — APPROVED** [review]
LGTM. Builds & tests pass!

File: src/client/gui/ffi/dart_ffi.cpp [hunk 1]
```diff
@@ -214,4 +215,23 @@ catch (...)
     mpl::log(mpl::Level::warning, category, "failed converting memory to bytes");
     return -1;
 }
+
+const char* default_mount_target(const char* source)
+try
+{
+    const QString q_source{source};
+    free((void*)source);
```
Comment (ricab) [inline src/client/gui/ffi/dart_ffi.cpp:223]: I am a little surprised that this takes ownership of the parameter and deletes memory that is supposed to hold const chars. I suppose the intention is that the client doesn't have to free after each function call, but in that case, don't you think we should remove the `const`? As it stands, it is easy to get into UB (e.g. call with a compile-time constant).

I realize this pattern is repeated all over this file, but it should be straightforward to modify everywhere (provided it works).

File: src/client/gui/ffi/dart_ffi.cpp [hunk 2]
```diff
@@ -214,4 +215,23 @@ catch (...)
...
+    return strdup(target.c_str());
+}
+catch (const std::exception& e)
+{
+    mpl::log(mpl::Level::warning, category, fmt::format("failed retrieving default mount target: {}", e.what()));
+    return nullptr;
+}
+catch (...)
+{
+    mpl::log(mpl::Level::warning, category, "failed retrieving default mount target");
```
Comment (ricab) [inline src/client/gui/ffi/dart_ffi.cpp:234]: Could you please factor out the error message? Can't use function try any longer, but that should be fine. You can have a `static constexpr char *` at the beginning.

File: include/multipass/dart_ffi.h [hunk 3]
```diff
@@ -40,6 +40,8 @@ int gid();
 int default_id();
 
 long long memory_in_bytes(const char* value);
+
+const char* default_mount_target(const char* source);
```
Comment (ricab) [inline include/multipass/dart_ffi.h:44]: Here too. This transfers ownership, so the `const` doesn't look right.

**Review by ricab — COMMENTED** [review]
Hey @andrei-toterman, thanks for modularising it in this way, it is great that we don't need to repeat the logic. I have a couple of matters for discussion below.

File: src/client/gui/ffi/dart_ffi.cpp (hunk 1 above)
Comment (andrei-toterman) [reply src/client/gui/ffi/dart_ffi.cpp:223]: That is a very good point, thanks. That is indeed a critical mistake. I'll fix it everywhere ASAP. I'll make a new PR for fixing those issues and set that PR as the base for this one. Is that ok?

File: src/client/gui/ffi/dart_ffi.cpp (hunk 2 above)
Comment (andrei-toterman) [reply src/client/gui/ffi/dart_ffi.cpp:234]: Yes, sure. It just sucks that it adds another level of indentation for the whole body of the function...

File: src/client/gui/ffi/dart_ffi.cpp (hunk 1 above)
Comment (ricab) [reply src/client/gui/ffi/dart_ffi.cpp:223]: Sure, that sounds perfectly fine :smiley:

**Review by ricab — DISMISSED** [review]
LGTM, thanks!

**Review by ricab — APPROVED** [review]
Merging this to main directly to speed things up.

**Pull request merged**